File: python/0020_valid_parentheses.py
# Time: O(n)
# Spase: O(n)


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        map = {")": "(", "}": "{", "]": "["}

        for char in s:
            if char in map.keys():
                if not stack or stack.pop() != map[char]:
                    return False
            else:
                stack.append(char)

        return not stack

File: go/0020_valid_parentheses.go
package main

// n = len(s)
// Time:  O(n)
// Space: O(n)
func isValid(s string) bool {
	brackets_pair := map[rune]rune{
		'}': '{',
		')': '(',
		']': '[',
	}
	stack := []rune{}

	for _, char := range s {
		if bracket, found := brackets_pair[char]; found {
			if len(stack) > 0 && stack[len(stack)-1] == bracket {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		} else {
			stack = append(stack, char)
		}
	}

	return len(stack) == 0
}
